enhancements/microshift/microshift-greenboot.md

```diff
@@ -0,0 +1,350 @@
+- How to implement upgrade fallbacks without using `greenboot`
+- How to upgrade Edge Devices, including best practices, etc.
+- How to monitor Edge Device software version and upgrade status (see USHIFT-497)
+- How to implement MicroShift data backup and restore for fallbacks (see ETCD-353 and USHIFT-518)
+- How to implement successful 3rd party software installation and upgrades
+- How to verify 3rd party software operation and provide a fallback if failed
+
+## Proposal
+
+### Workflow Description
+
+**Device Administrator** is a human user responsible for managing Edge Devices.
+
+**End User** is a human user operating an Edge Device.
+
+1. Device Administrator releases an updated system image over the network to
+upgrade devices in the field
+2. Edge Devices monitor system upgrades and download them in the background
+3. Edge Device upgrade staged on `rpm-ostree` does not take place until the next
+system reboot
+
+> In step 3, the system reboot may be scheduled by a Device Administrator, or
+> spontaneous due to a End User action, power failure, etc.
+
+1. End Users continue using the current software version until their devices are
+rebooted
+2. Upon a device reboot, End Users get a new software version only if it has been
+downloaded and staged
+3. End Users wait until the new software verification procedure completes
+4. Depending on the verification results, End Users get the updated software,
+or the upgrade is rolled back
+
+> In step 3, the verification procedure may reboot a device multiple times until
+> the upgrade is declared successful or failed.
+
+Once the system is rebooted implicitly (i.e. `cron` schedule) or explicitly
+by a Device Administrator, End User, etc.,
+the [greenboot-healthcheck.service](https://github.com/fedora-iot/greenboot#how-does-it-work)
+runs to verify the system validity.
+
+If the system validity procedures are successful, the version of the software
+is declared valid for end users. Otherwise, `greenboot` enters a system
+verification loop by repeating reboots up-to `GREENBOOT_MAX_BOOT_ATTEMPTS` times
+before attempting to rollback to the previous version of the software.
+
+```mermaid
+graph TD
+    A[Software vN] --> |Software Upgrade| B[Reboot Pending]
+    B --> |System Reboot| D{Validate Software vN+1}
+    D --> |Valid| E[Software vN+1]
+    D --> |Invalid| F[Verification Loop]
+    F --> |Up to Max Reboots|D
+    F --> |All Failed Attempts| G[Rollback]
+    G --> A
+```
+
+### API Extensions
+
+None
+
+### Implementation Details
+
+The MicroShift core RPM package will include the `microshift-greenboot-required.sh`
+health check script to validate that all the required MicroShift services are
+up and running. There will be no explicit dependency on the `greenboot` RPM.
+The script will be installed into the `/etc/greenboot/check/required.d` directory
+and it will not be executed during the system boot in case the `greenboot` package
```

Comment:
On R4E, `greenboot` is always present. Also, even if it is present, there may be corner cases where the user considers a MicroShift health check failing not a criterion for rollback.
Why not bundle the script in a `microshift-greenboot` RPM that may or may not be installed by the user?

Reply:
For the `greenboot` being optional, we also support running MicroShift on a regular RHEL where greenboot may not be present. Even on RHDE OS, it can be removed by `rpm-ostree override remove greenboot greenboot-default-health-checks`

For a separate RPM, it is just one script. Should we create RPM for it? We may decide to do it eventually, but there does not seem to be a strong case for it right now, IMO.

Reply:
Done